### PR TITLE
rac2,tracker: use CircularBuffer for tracker.Inflights

### LINKED ISSUE #135223: raft: PANIC inflights buffer is never resized beyond the initial size
Seen in this OOB panic:

```

I241114 21:22:22.551442 12898 kv/kvserver/replica_raftstorage.go:521 ⋮ [T1,Vsystem,n2,s3,r35/9:‹/Table/3{2-3}›] 291  applied snapshot df45e38e from (n9,s17):7 at applied index 74 ‹as write ›(total=1ms data=2.2 KiB excise=true ingestion=6@1ms)
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292  a panic has occurred!
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +runtime error: index out of range [128] with length 128
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +(1) attached stack trace
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  -- stack trace:
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | runtime.gopanic
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       GOROOT/src/runtime/panic.go:770
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | runtime.goPanicIndex
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       GOROOT/src/runtime/panic.go:114
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/raft/tracker.(*Inflights).Add
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/raft/tracker/inflights.go:84
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/raft/tracker.(*Progress).SentEntries
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/raft/tracker/progress.go:191
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/raft.(*raft).prepareMsgApp
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/raft/raft.go:625
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/raft.(*raft).maybePrepareMsgApp
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/raft/raft.go:654
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/raft.(*RawNode).SendMsgApp
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/raft/rawnode.go:210
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.raftNodeForRACv2.SendMsgAppRaftMuLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/kvflowcontrol/replica_rac2/raft_node.go:74
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.(*replicaSendStream).handleReadyEntriesRaftMuAndStreamLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go:2651
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.(*replicaState).handleReadyEntriesRaftMuLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go:2295
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.(*rangeController).HandleRaftEventRaftMuLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go:1172
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*processorImpl).HandleRaftReadyRaftMuLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go:784
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/replica_raft.go:1017
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/replica_raft.go:836
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/store_raft.go:682
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/scheduler.go:419
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/kv/kvserver/scheduler.go:319
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       pkg/util/stop/stopper.go:498
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  | runtime.goexit
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +  |       src/runtime/asm_amd64.s:1695
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +Wraps: (2) runtime error: index out of range [128] with length 128
E241114 21:22:34.036920 406 1@util/log/logcrash/crash_reporting.go:192 ⋮ [T1,Vsystem,n2] 292 +Error types: (1) *withstack.withStack (2) runtime.boundsError
```

Which is due to the grow function not actually allowing growth beyond `Inflights.size`:

https://github.com/cockroachdb/cockroach/blob/ef26d1534507f6089aa909e0bd43a0efb71dd6c7/pkg/raft/tracker/inflights.go#L96-L98

This affects master and v24.3. It has only reproduced with `perturbation/full/backfill` with `appply_to_all` set for `kvadmission.flow_control.mode` and the elastic disk BW limiter enabled.

Tentatively marking as a GA blocker. The fix appears simple.

Jira issue: CRDB-44405

### PR DESCRIPTION
**Pull request by sumeerbhola**
The custom circular/ring buffer implementation in Raft's tracker.Inflights does not shrink and unnecessarily grows to the maximum configured size (unless the buffer becomes empty). With RAC2, the maximum size is no longer a maximum, in lazy replication mode. There was a bug in incorrectly trying to impose this maximum, which motivated this slightly bigger cleanup. In general, it is beneficial for the inflight buffer to also shrink, and by using the existing CircularBuffer, we can simplify the code in tracker.Inflights.

As part of this change rac2.CircularBuffer is moved to util/container/ring and becomes ring.Buffer. There is another ring.Buffer in util/ring, which doesn't shrink and is not optimized to use bit arithmetic -- there is a todo added there to replace it with the implementation in util/container/ring.

Fixes #135223

Epic: none

Release note: None

### CODE REVIEW

**Review by kvoli — COMMENTED** [review]
Looks good, I'll cherry-pick this SHA and give it a run on some roachtests.

> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-:-OBhTcyJANUszg1VnDLc:b-an70rn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/raft/tracker/inflights.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhQ3J3-YOofoFTx65-:-OBhQ3J3-YOofoFTx650:b7jvusw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/raft/tracker/inflights.go#L38)):*
> ```Go
> 	// TODO(pav-kv): do not store the limits here, pass them to methods. For flow
> 	// control, we need to support dynamic limits.
> 	size     int    // the max number of inflight messages
> ```

Should `size` be removed as well? EDIT: ignore.
___
*[`pkg/raft/tracker/inflights.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhQZOmEWg7QZdz8ToT:-OBhQZOmEWg7QZdz8ToU:b-rirk6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/raft/tracker/inflights.go#L102)):*
> ```Go
> // Full returns true if no more messages can be sent at the moment.
> func (in *Inflights) Full() bool {
> 	return in.buffer.Length() >= in.size || (in.maxBytes != 0 && in.bytes >= in.maxBytes)
> ```

I was about to comment re: removing `size` but then saw this. Is this only necessary for compatibility w/ push mode? Presumably, there is no such thing as the Inflights being `Full()` on the number of entries in pull mode, only the `in.bytes >= in.maxBytes` conditional?
___
*[`pkg/util/container/ring/buffer.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhS-z68y_ZJPoP_34E:-OBhS-z68y_ZJPoP_34F:b-dziupf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/util/container/ring/buffer.go#L37)):*
> ```Go
> // is arbitrary.
> //
> // TODO: should we reduce this?
> ```

I would, or allow a capacity to be provided at init. That said, some quick math for the increase in mem usage with some assumptions:

```
entry_size                           = 2 * uint64 = 16 bytes
replicas                               = 20k (upper bound really)
b4_buffer_size_avg_len     = 8
now_buffer_size_avg_len  = 32
                                  delta  = 24

extra_mem = 24 * 16 * 20000 / (1<<20)
                     = 7.3 MiB
```

So it doesn't cost that much, assuming my numbers aren't way off.

**Review by kvoli — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-:-OBhXCY45paAYcIt3K_0:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

**Review by sumeerbhola — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-:-OBhh7TK-J95T0DxuRmr:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @kvoli and @pav-kv)

___
*[`pkg/raft/tracker/inflights.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhQ3J3-YOofoFTx65-:-OBhdEf27PkKm6z1CxcW:b-ahhm9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/raft/tracker/inflights.go#L38)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Should `size` be removed as well? I
</blockquote></details>

It's really a limit for eager replication mode, so we need it. It should be renamed though. We can leave that renaming etc. for the existing TODO.
___
*[`pkg/raft/tracker/inflights.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhQZOmEWg7QZdz8ToT:-OBhdZpn1dXPer_HjdoR:b-8luy4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/raft/tracker/inflights.go#L102)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I was about to comment re: removing `size` but then saw this. Is this only necessary for compatibility w/ push mode? Presumably, there is no such thing as the Inflights being `Full()` on the number of entries in pull mode, only the `in.bytes >= in.maxBytes` conditional?
</blockquote></details>

Yes, we need it for eager replication/push mode. Even the `maxBytes` is irrelevant in pull mode. This `Full` method only affects control in `CanSendEntries` and `IsPaused`. The latter I was asking about on Slack. I think those calls are harmless, but there seems to be some scope for bugs without careful documentation.
___
*[`pkg/util/container/ring/buffer.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/135237#-OBhS-z68y_ZJPoP_34E:-OBheNg-9lAygtGnLxMT:b-u6tnkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9e718cc5c9d64f059df2f20663d68e25a154645/pkg/util/container/ring/buffer.go#L37)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I would, or allow a capacity to be provided at init. That said, some quick math for the increase in mem usage with some assumptions:

```
entry_size                           = 2 * uint64 = 16 bytes
replicas                               = 20k (upper bound really)
b4_buffer_size_avg_len     = 8
now_buffer_size_avg_len  = 32
                                  delta  = 24

extra_mem = 24 * 16 * 20000 / (1<<20)
                     = 7.3 MiB
```

So it doesn't cost that much, assuming my numbers aren't way off.
</blockquote></details>

Thanks for the calculation. I removed this TODO. We can add an initial capacity if needed in the future.

File: pkg/kv/kvserver/kvflowcontrol/rac2/log_tracker.go [hunk 1]
```diff
@@ -109,7 +110,7 @@ type LogTracker struct {
 	//	- waiting[pri][i].Term <= last.Term
 	//	- waiting[pri][i].Index < waiting[pri][i+1].Index
 	//	- waiting[pri][i].Term <= waiting[pri][i+1].Term
-	waiting [raftpb.NumPriorities]CircularBuffer[LogMark]
```
Comment (tbg) [inline pkg/kv/kvserver/kvflowcontrol/rac2/log_tracker.go:112]: just a drive-by that you don't have to address but something for "next time": I would prefer to see this in separate commits. As is, the commit tangles extracting the `ring.Buffer` with adopting it in `RawNode`.

File: pkg/raft/tracker/inflights_test.go [hunk 2]
```diff
@@ -112,76 +94,61 @@ func TestInflightFreeTo(t *testing.T) {
...
 	wantIn0 := &Inflights{
-		start: 1,
-		count: 9,
 		bytes: 945,
 		size:  10,
 		buffer: inflightsBuffer(
-			//          ↓------------------------
-			[]uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
-			[]uint64{100, 101, 102, 103, 104, 105, 106, 107, 108, 109}),
+			//       ↓------------------------
```
Comment (pav-kv) [inline pkg/raft/tracker/inflights_test.go:100]: Can drop these pointer comments now since they all point at the first entry. Same below and above.

**Review by pav-kv — APPROVED** [review]
LGTM, thank you. One nit, but feel free to defer if it delays merge.

File: pkg/raft/tracker/inflights_test.go (hunk 2 above)
Comment (sumeerbhola) [reply pkg/raft/tracker/inflights_test.go:100]: will do in a followup. TFTR!

**Comment by sumeerbhola** [comment]
bors r=kvoli,pav-kv

**Pull request merged**